File: weather-demo-app-android/app/src/main/java/com/demo/weather/extension/WeatherExtension.kt
package com.demo.weather.extension

import com.demo.weather.domain.dto.CurrentWeatherDto
import com.demo.weather.domain.dto.ForecastDto
import com.demo.weather.domain.dto.WeatherInfoDto
import com.demo.weather.domain.model.CurrentWeather
import com.demo.weather.domain.model.ForecastDay
import com.demo.weather.domain.model.WeatherCondition

fun CurrentWeatherDto.toDomain(): CurrentWeather {
    val firstWeather = weather.firstOrNull()

    return CurrentWeather(
        city = name,
        condition = firstWeather.toCondition(),
        description = firstWeather?.description.orEmpty(),
        temperatureCelsius = main.temp,
        feelsLikeCelsius = main.feelsLike,
        humidityPercent = main.humidity,
        windSpeedKph = wind.speed * 3.6,
        iconCode = firstWeather?.icon.orEmpty()
    )
}

fun ForecastDto.toDomainDaily(): List<ForecastDay> {
    return list
        .groupBy { it.dateText.substringBefore(" ") }
        .mapNotNull { (date, items) ->
            val midday = items.firstOrNull { it.dateText.contains("12:00:00") } ?: items.firstOrNull()
            val firstWeather = midday?.weather?.firstOrNull()

            midday?.let {
                ForecastDay(
                    date = date,
                    condition = firstWeather.toCondition(),
                    description = firstWeather?.description.orEmpty(),
                    temperatureCelsius = it.main.temp,
                    feelsLikeCelsius = it.main.feelsLike,
                    iconCode = firstWeather?.icon.orEmpty()
                )
            }
        }
        .take(5)
}

private fun WeatherInfoDto?.toCondition(): WeatherCondition {
    return when (this?.main?.lowercase()) {
        "clear" -> WeatherCondition.SUNNY
        "clouds" -> WeatherCondition.CLOUDY
        "rain", "drizzle" -> WeatherCondition.RAIN
        "snow" -> WeatherCondition.SNOW
        "thunderstorm" -> WeatherCondition.THUNDERSTORM
        "mist", "fog", "haze", "smoke" -> WeatherCondition.MIST
        else -> WeatherCondition.UNKNOWN
    }
}

File: weather-demo-app-apple/Weather/Extension/WeatherExtension.swift
//
//  WeatherMapper.swift
//  Weather
//
//  Created by Onur on 4/11/26.
//

import Foundation

extension CurrentWeatherDTO {
    func toDomain() -> CurrentWeather {
        let firstWeather = weather.first

        return CurrentWeather(
            city: name,
            condition: firstWeather.toCondition(),
            description: firstWeather?.description ?? "",
            temperatureCelsius: main.temp,
            feelsLikeCelsius: main.feelsLike,
            humidityPercent: main.humidity,
            windSpeedKph: wind.speed * 3.6,
            iconCode: firstWeather?.icon ?? ""
        )
    }
}

extension ForecastDTO {
    func toDomainDaily() -> [ForecastDay] {
        let grouped = Dictionary(grouping: list) { item in
            item.dateText.components(separatedBy: " ").first ?? item.dateText
        }

        return grouped
            .sorted { $0.key < $1.key }
            .compactMap { date, items in
                let midday = items.first(where: { $0.dateText.contains("12:00:00") }) ?? items.first
                let firstWeather = midday?.weather.first

                guard let midday else { return nil }

                return ForecastDay(
                    date: date,
                    condition: firstWeather.toCondition(),
                    description: firstWeather?.description ?? "",
                    temperatureCelsius: midday.main.temp,
                    feelsLikeCelsius: midday.main.feelsLike,
                    iconCode: firstWeather?.icon ?? ""
                )
            }
            .prefix(5)
            .map { $0 }
    }
}

private extension Optional where Wrapped == WeatherInfoDTO {
    func toCondition() -> WeatherCondition {
        switch self?.main.lowercased() {
        case "clear":
            return .sunny
        case "clouds":
            return .cloudy
        case "rain", "drizzle":
            return .rain
        case "snow":
            return .snow
        case "thunderstorm":
            return .thunderstorm
        case "mist", "fog", "haze", "smoke":
            return .mist
        default:
            return .unknown
        }
    }
}
